modules/tree/main/index.coffee

```diff
@@ -10,33 +40,44 @@ formatIcon = (node, iconElement, animate) =>
         .go(true)
 
 format = (tree, element, animate) ->
+  treeNode = hx.select(element)
+  treeNodeComponent = treeNode.component()
+
+  renderLazyChildren = (rootNode, recursive) ->
+    if rootNode.component()? and rootNode.selectAll('.hx-tree-node').empty()
+      rootNode.component().renderChildren()
+      formatChildren(tree, rootNode.selectAll('.hx-tree-node'), animate)
+    if recursive
+      rootNode.selectAll('.hx-tree-node').map((sel) -> renderLazyChildren(sel, recursive))
+
   toggle = (iconElement) =>
-    selection = hx.select(element).select('.hx-tree-node-children')
+    selection = treeNode.select('.hx-tree-node-children')
+    renderLazyChildren(treeNode)
     display = if selection.style('display') is 'none' then 'block' else 'none'
     selection.style('display', display)
     formatIcon(element, iconElement, tree.options.animate)
 
   # tries to open all children of the target toggle if it is already open, or if the node is closed, it simply opens it
   openAllOrToggle = (iconElement) =>
-    selection = hx.select(element).selectAll('.hx-tree-node-children')
-    root = hx.select(element).select('.hx-tree-node-children')
-    rootNode = root.node()
-
-    if root.style('display') is 'block'
-      selection.forEach (node) ->
-        if node isnt rootNode
-          if node.style('display') isnt 'block'
-            parentNode = node.node().parentNode
-            iconElement = hx.select(parentNode).select('.hx-tree-node-parent-icon').node()
-            node.style('display', 'block')
-            formatIcon(parentNode, iconElement, tree.options.animate)
+    renderLazyChildren(treeNode, true)
+    treeNode.selectAll('.hx-tree-node-children').style('display', 'block')
+
+    if treeNode.style('display') is 'block'
+      formatChildren(tree, treeNode.selectAll('.hx-tree-node'), animate)
```

Comment:
Thinking about it, the reason this is slow is because it's using a `view` to render the children - Could we instead change `view` to only render everything before appending? (if it doesn't already do that)